File: app/src/main/java/com/twokwy/chessandroid/ChessBoardFactory.java
package com.twokwy.chessandroid;

import android.content.Context;
import android.graphics.Rect;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class ChessBoardFactory {
    private final int darkSquareColor;
    private final int lightSquareColor;

    private static final char[] FILES = {'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'};
    private static final int[] RANKS = {1, 2, 3, 4, 5, 6, 7, 8};

    public ChessBoardFactory(Context context) {
        darkSquareColor = context.getResources().getColor(R.color.dark_square);
        lightSquareColor = context.getResources().getColor(R.color.light_square);
    }

    /**
     * Creates a chess board with 8x8 squares per to fit into the given dimensions.
     *
     * @param w width (px) available for the chess board
     * @param h height (px) available for the chess board
     */
    public ChessBoard createToFillWidthAndHeight(int w, int h) {
        final int maxSquareHeight = h / 8;
        final int maxSquareWidth = w / 8;

        final int squareSize = Math.min(maxSquareHeight, maxSquareWidth);

        int xOffset = (w - (squareSize * 8)) / 2;
        int yOffset = (h - (squareSize * 8)) / 2;

        List<ChessSquare> squares = createListOfSquaresForBoard(squareSize, xOffset, yOffset);

        return new ChessBoard(squares);
    }

    private List<ChessSquare> createListOfSquaresForBoard(
            int squareSize, int xOffset, int yOffset) {
        List<ChessSquare> squares = new ArrayList<>(64);
        for (int i = 0; i < 64; i++) {
            int x = i % 8;
            int y = i / 8;
            char file = FILES[x];
            int rank = RANKS[7 - y];
            String location = "" + file + rank;

            int leftBound = xOffset + x * squareSize;
            int rightBound = leftBound + squareSize;
            int topBound = yOffset + y * squareSize;
            int bottomBound = topBound + squareSize;

            Rect bounds = new Rect(leftBound, topBound, rightBound, bottomBound);
            int color = (x + y) % 2 == 0 ? darkSquareColor : lightSquareColor;
            ChessPiece chessPiece = getStartingPieceAt(x, y);
            squares.add(
                    i,
                    ChessSquare.create(location, color, bounds, Optional.ofNullable(chessPiece)));
        }
        return squares;
    }

    private static ChessPiece getStartingPieceAt(int x, int y) {
        ChessPieceType pieceType = getChessPieceTypeAt(x, y);
        if (pieceType == null) {
            return null;
        }
        return new ChessPiece(pieceType, y < 2 ? PieceColor.BLACK : PieceColor.WHITE);
    }

    private static ChessPieceType getChessPieceTypeAt(int x, int y) {
        if (y == 1 || y == 6) {
            return ChessPieceType.PAWN;
        } else if ((y == 0 || y == 7)) {
            switch (x) {
                case 0:
                case 7:
                    return ChessPieceType.ROOK;
                case 1:
                case 6:
                    return ChessPieceType.KNIGHT;
                case 2:
                case 5:
                    return ChessPieceType.BISHOP;
                case 3:
                    return ChessPieceType.KING;
                case 4:
                    return ChessPieceType.QUEEN;
            }
        }
        return null;
    }
}

File: app/src/main/java/com/twokwy/chessandroid/board/ChessBoardFactory.kt
package com.twokwy.chessandroid.board

import android.content.Context
import android.graphics.Rect
import com.twokwy.chessandroid.R
import com.twokwy.chessandroid.icons.ChessPieceIconsImpl
import com.twokwy.chessandroid.pieces.ChessPiece
import com.twokwy.chessandroid.pieces.ChessPieceType
import com.twokwy.chessandroid.pieces.PieceColor

class ChessBoardFactory(private val context: Context) {
    private val darkSquareColor: Int = context.resources.getColor(R.color.dark_square)
    private val lightSquareColor: Int = context.resources.getColor(R.color.light_square)

    /**
     * Creates a chess board with 8x8 squares per to fit into the given dimensions.
     *
     * @param w width (px) available for the chess board
     * @param h height (px) available for the chess board
     */
    fun createToFillWidthAndHeight(w: Int, h: Int): ChessBoard {
        val maxSquareHeight = h / 8
        val maxSquareWidth = w / 8
        val squareSize = Math.min(maxSquareHeight, maxSquareWidth)
        val xOffset = (w - squareSize * 8) / 2
        val yOffset = (h - squareSize * 8) / 2
        val squares = createListOfSquaresForBoard(squareSize, xOffset, yOffset)
        return ChessBoard(squares, ChessPieceIconsImpl(context))
    }

    private fun createListOfSquaresForBoard(
            squareSize: Int, xOffset: Int, yOffset: Int): List<ChessSquare> {
        val squares: MutableList<ChessSquare> = ArrayList(64)
        for (i in 0..63) {

            val x = i % 8
            val y = i / 8
            val location = toLocation(x, y)

            val leftBound = xOffset + x * squareSize
            val rightBound = leftBound + squareSize
            val topBound = yOffset + y * squareSize
            val bottomBound = topBound + squareSize
            val bounds = Rect(leftBound, topBound, rightBound, bottomBound)
            val color = if ((x + y) % 2 == 0) darkSquareColor else lightSquareColor
            val chessPiece = getStartingPieceAt(x, y)
            squares.add(
                    i,
                    ChessSquare.create(location, bounds, chessPiece, color))
        }
        return squares
    }

    private fun toLocation(x: Int, y: Int): Location = Location(x, y)

    companion object {
        private fun getStartingPieceAt(x: Int, y: Int): ChessPiece? {
            val pieceType = getChessPieceTypeAt(x, y) ?: return null
            return ChessPiece(pieceType, if (y < 2) PieceColor.BLACK else PieceColor.WHITE)
        }

        private fun getChessPieceTypeAt(x: Int, y: Int): ChessPieceType? {
            if (y == 1 || y == 6) {
                return ChessPieceType.PAWN
            } else if (y == 0 || y == 7) {
                when (x) {
                    0, 7 -> return ChessPieceType.ROOK
                    1, 6 -> return ChessPieceType.KNIGHT
                    2, 5 -> return ChessPieceType.BISHOP
                    3 -> return ChessPieceType.KING
                    4 -> return ChessPieceType.QUEEN
                }
            }
            return null
        }
    }
}
